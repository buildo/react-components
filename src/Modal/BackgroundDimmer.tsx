import * as React from 'react';
import * as cx from 'classnames';
import FlexView from 'react-flexview';
import omit = require('lodash/omit');
import { findDOMNode, Children } from '../utils';

export type BackgroundDimmerRequiredProps = {
  /** children nodes/elements */
  children: Children;
  /** called when user clicks outside children */
  onClickOutside?: (e: React.SyntheticEvent<HTMLDivElement>) => void;
  /** component's class */
  className?: string;
  /** component's id */
  id?: string;
  /** component's style */
  style?: React.CSSProperties;
};

export type BackgroundDimmerDefaultProps = {
  /** background-color */
  color: string;
  /** opacity */
  alpha: number;
  /** z-index (BackgroundDimmer has `position: fixed`) */
  zIndex: number;
  /** centeredContentWrapper width */
  width: string | number;
  /** centeredContentWrapper max-width */
  maxWidth: string | number;
  /** centeredContentWrapper height */
  height: string | number;
  /** centeredContentWrapper max-height */
  maxHeight: string | number;
  /** avoid propagation for scroll events */
  stopScrollPropagation: boolean;
};

type BackgroundDimmerDefaultedProps = BackgroundDimmerRequiredProps & BackgroundDimmerDefaultProps;

export namespace BackgroundDimmer {
  export type Props = BackgroundDimmerRequiredProps & Partial<BackgroundDimmerDefaultProps>;
}

export class BackgroundDimmer extends React.PureComponent<BackgroundDimmer.Props> {
  private mainContentWrapper: FlexView | null = null;

  static defaultProps: BackgroundDimmerDefaultProps = {
    color: 'black',
    alpha: 0.5,
    zIndex: 99999,
    width: 'auto',
    maxWidth: '90%',
    height: 'auto',
    maxHeight: '90%',
    stopScrollPropagation: false
  };

  // srcElement is not declared in React.SyntheticEvent cause it's a key used by old
  // IE versions: https://msdn.microsoft.com/en-us/library/ms534638(v=vs.85).aspx
  // TODO: should be removed in the future
  isEventOutsideChildren = (
    e: React.SyntheticEvent<HTMLDivElement> & { srcElement?: EventTarget }
  ) => {
    const el = e.target || e.srcElement;
    return this.mainContentWrapper && el === findDOMNode(this.mainContentWrapper);
  };

  onClick = (e: React.SyntheticEvent<HTMLDivElement>) => {
    const { onClickOutside } = this.props;
    if (onClickOutside) {
      onClickOutside(e);
    }
  };

  stopPropagation = (e: React.SyntheticEvent<HTMLDivElement>) => e.stopPropagation();

  preventDefault = (e: React.SyntheticEvent<HTMLDivElement>) => e.preventDefault();

  stopScrollPropagation = (e: React.SyntheticEvent<HTMLDivElement>) => {
    if (this.props.stopScrollPropagation && this.isEventOutsideChildren(e)) {
      this.preventDefault(e);
    }
  };

  render() {
    const {
      className,
      zIndex,
      color,
      alpha: opacity,
      width,
      maxWidth,
      height,
      maxHeight,
      children,
      ...props
    } = this.props as BackgroundDimmerDefaultedProps;
    const { onClick, stopPropagation, stopScrollPropagation } = this;

    const fixedStyle: React.CSSProperties = {
      position: 'fixed',
      top: 0,
      left: 0,
      right: 0,
      bottom: 0
    };
    const overlayProps = {
      style: {
        ...fixedStyle,
        zIndex,
        backgroundColor: color,
        opacity
      }
    };
    const mainContentWrapperProps: FlexView.Props & {
      ref: (ref: FlexView | null) => void;
    } = {
      onClick,
      onWheel: stopScrollPropagation,
      onTouchMove: stopScrollPropagation,
      style: { ...fixedStyle, zIndex: zIndex + 1 },
      className: 'main-content-wrapper',
      vAlignContent: 'center',
      hAlignContent: 'center',
      ref: ref => {
        this.mainContentWrapper = ref;
      }
    };
    const centeredContentWrapperProps = {
      className: 'centered-content-wrapper',
      style: { width, maxWidth, height, maxHeight },
      onClick: stopPropagation,
      column: true
    };

    return (
      <div
        className={cx('background-dimmer', className)}
        {...omit(props, ['onClickOutside', 'stopScrollPropagation'])}
      >
        <div {...overlayProps} />
        <FlexView {...mainContentWrapperProps}>
          <FlexView {...centeredContentWrapperProps}>{children}</FlexView>
        </FlexView>
      </div>
    );
  }
}
