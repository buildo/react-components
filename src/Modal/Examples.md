### Examples

```js
intialState = { isOpen: false };

function open() {
  setState({ isOpen: true });
}

function close() {
  setState({ isOpen: false });
}

const closeIcon = (
  <svg version="1.1" xmlns="http://www.w3.org/2000/svg" height="12" viewBox="0 0 22 28">
    <path
      fill="#9098a7"
      d="M20.281 20.656q0 0.625-0.438 1.062l-2.125 2.125q-0.438 0.438-1.062 0.438t-1.062-0.438l-4.594-4.594-4.594 4.594q-0.438 0.438-1.062 0.438t-1.062-0.438l-2.125-2.125q-0.438-0.438-0.438-1.062t0.438-1.062l4.594-4.594-4.594-4.594q-0.438-0.438-0.438-1.062t0.438-1.062l2.125-2.125q0.438-0.438 1.062-0.438t1.062 0.438l4.594 4.594 4.594-4.594q0.438-0.438 1.062-0.438t1.062 0.438l2.125 2.125q0.438 0.438 0.438 1.062t-0.438 1.062l-4.594 4.594 4.594 4.594q0.438 0.438 0.438 1.062z"
    ></path>
  </svg>
);

const getModal = () => (
  <Modal
    transitionEnterTimeout={500}
    transitionLeaveTimeout={500}
    onDismiss={close}
    iconClose={closeIcon}
    title="Send Message"
    footer={
      <FlexView hAlignContent="right">
        <Button size="small" style={{ marginRight: 10 }} onClick={close}>
          Cancel
        </Button>
        <Button primary size="small" onClick={close}>
          Confirm
        </Button>
      </FlexView>
    }
  >
    Are you sure you want to send this message?
  </Modal>
);

<div>
  <Button onClick={open}>Open modal</Button>
  {state.isOpen && getModal()}
</div>;
```

#### Modal without header

```js
intialState = { isOpen: false };

function open() {
  setState({ isOpen: true });
}

function close() {
  setState({ isOpen: false });
}

const getModal = () => (
  <Modal
    transitionEnterTimeout={500}
    transitionLeaveTimeout={500}
    onDismiss={close}
    footer={
      <FlexView hAlignContent="right">
        <Button size="small" style={{ marginRight: 10 }} onClick={close}>
          Cancel
        </Button>
        <Button primary size="small" onClick={close}>
          Confirm
        </Button>
      </FlexView>
    }
  >
    Are you sure you want to send this message?
  </Modal>
);

<div>
  <Button onClick={open}>Open modal</Button>
  {state.isOpen && getModal()}
</div>;
```

#### Modal without footer

```js
intialState = { isOpen: false };

function open() {
  setState({ isOpen: true });
}

function close() {
  setState({ isOpen: false });
}

const closeIcon = (
  <svg version="1.1" xmlns="http://www.w3.org/2000/svg" height="12" viewBox="0 0 22 28">
    <path
      fill="#9098a7"
      d="M20.281 20.656q0 0.625-0.438 1.062l-2.125 2.125q-0.438 0.438-1.062 0.438t-1.062-0.438l-4.594-4.594-4.594 4.594q-0.438 0.438-1.062 0.438t-1.062-0.438l-2.125-2.125q-0.438-0.438-0.438-1.062t0.438-1.062l4.594-4.594-4.594-4.594q-0.438-0.438-0.438-1.062t0.438-1.062l2.125-2.125q0.438-0.438 1.062-0.438t1.062 0.438l4.594 4.594 4.594-4.594q0.438-0.438 1.062-0.438t1.062 0.438l2.125 2.125q0.438 0.438 0.438 1.062t-0.438 1.062l-4.594 4.594 4.594 4.594q0.438 0.438 0.438 1.062z"
    ></path>
  </svg>
);

const getModal = () => (
  <Modal
    transitionEnterTimeout={500}
    transitionLeaveTimeout={500}
    onDismiss={close}
    iconClose={closeIcon}
    title="Informative Modal"
  >
    This modal contains just info. So it is not possible to perform an action.
  </Modal>
);

<div>
  <Button onClick={open}>Open modal</Button>
  {state.isOpen && getModal()}
</div>;
```

#### Passing arbitrary React context

Modal renders through a "portal", so React context is not propagated as expected.

_We know context isn't a public api, but it is required for [widely](https://github.com/reactjs/react-redux) [used](https://github.com/yahoo/react-intl) libraries to work. We also know there's `renderSubtreeIntoContainer`, which is unfortunately even more unstable._

To solve this problem we export a handy `modalWithContext` that easily allows to pass context down correctly (look at the code!).

```js
const FooType = React.PropTypes.string.isRequired;
const closeIcon = (
  <svg version="1.1" xmlns="http://www.w3.org/2000/svg" height="12" viewBox="0 0 22 28">
    <path
      fill="#9098a7"
      d="M20.281 20.656q0 0.625-0.438 1.062l-2.125 2.125q-0.438 0.438-1.062 0.438t-1.062-0.438l-4.594-4.594-4.594 4.594q-0.438 0.438-1.062 0.438t-1.062-0.438l-2.125-2.125q-0.438-0.438-0.438-1.062t0.438-1.062l4.594-4.594-4.594-4.594q-0.438-0.438-0.438-1.062t0.438-1.062l2.125-2.125q0.438-0.438 1.062-0.438t1.062 0.438l4.594 4.594 4.594-4.594q0.438-0.438 1.062-0.438t1.062 0.438l2.125 2.125q0.438 0.438 0.438 1.062t-0.438 1.062l-4.594 4.594 4.594 4.594q0.438 0.438 0.438 1.062z"
    ></path>
  </svg>
);

const ModalWithContext = modalWithContext({ foo: FooType });

class ComponentAccessingContext extends React.Component {
  render() {
    return <div>{this.context.foo}</div>;
  }
}

ComponentAccessingContext.contextTypes = {
  foo: FooType
};

class Component extends React.Component {
  getChildContext() {
    return { foo: 'this comes from context!' };
  }

  constructor() {
    this.state = { isOpen: false };
    this.open = this.open.bind(this);
    this.close = this.close.bind(this);
  }

  open() {
    this.setState({ isOpen: true });
  }

  close() {
    this.setState({ isOpen: false });
  }

  getModal() {
    return (
      <ModalWithContext
        transitionEnterTimeout={500}
        transitionLeaveTimeout={500}
        onDismiss={this.close}
        iconClose={closeIcon}
        title="Send Message"
        footer={
          <FlexView hAlignContent="right">
            <Button size="small" style={{ marginRight: 10 }} onClick={this.close}>
              Cancel
            </Button>
            <Button primary size="small" onClick={this.close}>
              Confirm
            </Button>
          </FlexView>
        }
      >
        <ComponentAccessingContext />
      </ModalWithContext>
    );
  }

  render() {
    return (
      <div>
        <Button onClick={this.open}>Open modal</Button>
        {this.state.isOpen && this.getModal()}
      </div>
    );
  }
}

Component.childContextTypes = {
  foo: FooType
};

<Component />;
```
