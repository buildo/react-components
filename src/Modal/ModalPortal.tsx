import * as React from 'react';
import * as ReactDOM from 'react-dom';
import * as ReactTransitionGroup from 'react-transition-group/TransitionGroup';
import TransitionWrapper from '../TransitionWrapper';

let containerNode: Element | null = null;

export namespace ModalPortal {
  export type Props = {
    children: JSX.Element;
    /** transition enter timeout */
    transitionEnterTimeout: number;
    /** transition leave timeout */
    transitionLeaveTimeout: number;
    /** context types to pass to the modal React tree */
    childContextTypes?: React.ValidationMap<any>;
    /** should return context values to pass to the modal React tree */
    getChildContext?: () => object;
    className?: string;
  };
}

type ContextWrapperProps = {
  modal: () => JSX.Element;
  getChildContext?: () => object;
};

export class ModalPortal extends React.Component<ModalPortal.Props> {
  private isOpen: boolean = false;

  render() {
    return null;
  }

  _ContextWrapper: React.ComponentType<ContextWrapperProps> = (() => {
    const childContextTypes = this.props.childContextTypes || {};
    const getChildContext = this.props.getChildContext || (() => ({}));
    return class ContextWrapper extends React.Component<ContextWrapperProps> {
      // eslint-disable-line react/no-multi-comp
      static childContextTypes = childContextTypes;
      getChildContext = getChildContext;
      render() {
        return this.props.modal();
      }
    };
  })();

  componentDidMount() {
    this.isOpen = true;
    this._render();
  }

  componentDidUpdate() {
    this._render();
  }

  componentWillUnmount() {
    this.isOpen = false;
    this._render();
  }

  _cleanup() {
    if (containerNode) {
      ReactDOM.unmountComponentAtNode(containerNode);
      document.body.removeChild(containerNode);
      containerNode = null;
    }
  }

  _onLeave = () => setTimeout(this._cleanup); // setTimeout is needed to avoid conflicts with ReactTransitionGroup setState

  _render() {
    if (!containerNode) {
      containerNode = document.createElement('div');
      document.body.appendChild(containerNode);
    }

    const modal = this._renderModal();
    const ContextWrapper = this._ContextWrapper;
    ReactDOM.render(<ContextWrapper modal={modal} />, containerNode);
  }

  _renderModal() {
    const { className, children, transitionEnterTimeout, transitionLeaveTimeout } = this.props;

    return () => (
      <ReactTransitionGroup>
        {this.isOpen ? (
          <TransitionWrapper
            key={0}
            className={className}
            onLeave={this._onLeave}
            transitionEnterTimeout={transitionEnterTimeout}
            transitionLeaveTimeout={transitionLeaveTimeout}
          >
            {children}
          </TransitionWrapper>
        ) : null}
      </ReactTransitionGroup>
    );
  }
}
