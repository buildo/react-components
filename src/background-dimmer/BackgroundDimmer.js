import React from 'react';
import omit from 'lodash/object/omit';

const propTypes = {
  children: React.PropTypes.node.isRequired,
  color: React.PropTypes.string,
  alpha: React.PropTypes.number,
  stopScrollPropagation: React.PropTypes.bool,
  onScrollOutside: React.PropTypes.func,
  onClickOutside: React.PropTypes.func,
  className: React.PropTypes.string,
  id: React.PropTypes.string,
  style: React.PropTypes.object
};

const BackgroundDimmer = React.createClass({

  propTypes: propTypes,

  getDefaultProps() {
    return {
      color: 'black',
      alpha: 0.5
    };
  },

  componentDidMount() {
    this.addScrollListener();
  },

  addScrollListener() {
    const dimmedBackground = this.refs.dimmedBackground.getDOMNode();
    dimmedBackground.addEventListener('wheel', this.onScroll);
    dimmedBackground.addEventListener('touchmove', this.onScroll);
  },

  removeScrollListener() {
    const dimmedBackground = this.refs.dimmedBackground.getDOMNode();
    dimmedBackground.removeEventListener('wheel', this.onScroll);
    dimmedBackground.removeEventListener('touchmove', this.onScroll);
  },

  isEventOutsideChildren(e) {
    const el = e.target || e.srcElement;
    return el === this.refs.dimmedBackground.getDOMNode();
  },

  onClick(e) {
    const { onClickOutside } = this.props;
    if (onClickOutside && this.isEventOutsideChildren(e)) {
      onClickOutside(e);
    }
  },

  onScroll(e) {
    const { stopScrollPropagation, onScrollOutside } = this.props;
    if (this.isEventOutsideChildren(e)) {
      if (stopScrollPropagation) {
        e.preventDefault();
      }
      if (onScrollOutside) {
        onScrollOutside(e);
      }
    }
  },

  getDimmedBackground() {
    const { color, alpha } = this.props;
    const style = {
      position: 'fixed',
      top: 0,
      left: 0,
      right: 0,
      bottom: 0,
      backgroundColor: color,
      opacity: alpha + ''
    };
    return <div style={style} onClick={this.onClick} ref='dimmedBackground' />;
  },

  render() {
    const { style, className, id, children } = this.props;
    const other = omit(this.props, Object.keys(propTypes));
    const mergedStyle = {
      position: 'relative',
      ...style
    };
    return (
      <div {...{ style: mergedStyle, className, id, ...other }}>
        {this.getDimmedBackground()}
        <div style={{ visibility: 'hidden' }}>{children}</div>
        <div style={{ position: 'absolute', width: '100%', height: '100%', top: 0, left: 0 }}>
          {children}
        </div>
      </div>
    );
  },

  componentWillUnmount() {
    this.removeScrollListener();
  }

});

export default BackgroundDimmer;
