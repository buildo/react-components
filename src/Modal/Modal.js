import React from 'react';
import cx from 'classnames';
import t from 'tcomb';
import { props } from 'tcomb-react';
import { pure } from 'revenge';
import FlexView from '../flex/FlexView';
import Icon from '../Icon/Icon';
import BackgroundDimmer from '../background-dimmer';
import TransitionWrapper from '../transition-wrapper';

class ContextWrapper extends React.Component {
  getChildContext = () => this.props.context;
  render = () => this.props.modal();
}

@pure
@props({
  children: t.ReactNode,
  title: t.maybe(t.String),
  footer: t.maybe(t.ReactNode),
  closeIcon: t.maybe(t.Boolean),
  dismissOnClickOut: t.maybe(t.Boolean),
  onDismiss: t.maybe(t.Function),
  className: t.maybe(t.String),
  transitionEnterTimeout: t.Number,
  transitionLeaveTimeout: t.Number
})
export default class Modal extends React.Component { //eslint-disable-line react/no-multi-comp
  static defaultProps = {
    onDismiss: () => {},
    title: '',
    dismissOnClickOut: false,
    closeIcon: true
  };

  render() {
    return null;
  }

  componentDidMount() {
    this._render();
  }

  componentDidUpdate() {
    this._render();
  }

  componentWillUnmount() {
    this.removeModal = true;
    this._render();

    const { transitionLeaveTimeout } = this.props;
    setTimeout(() => this._cleanup(), transitionLeaveTimeout + 50);
  }

  _cleanup() {
    if (this.containerNode) {
      React.unmountComponentAtNode(this.containerNode);
      document.body.removeChild(this.containerNode);
      this.containerNode = null;
    }
  }

  _render() {
    if (!this.containerNode) {
      this.containerNode = document.createElement('div');
      document.body.appendChild(this.containerNode);
    }

    const Modal = this._renderModal();

    React.render(<ContextWrapper modal={Modal} />, this.containerNode);
  }

  _renderModal() {
    const {
      props: {
        children, title, closeIcon, className, footer, dismissOnClickOut,
        onDismiss,
        transitionEnterTimeout, transitionLeaveTimeout
      }
    } = this;
    const titleProps = {
      h3: {
        className: closeIcon ? 'with-close-icon' : null
      }
    };

    return () => (
      <React.addons.TransitionGroup transitionAppear>
        {!this.removeModal ? (
          <TransitionWrapper key={0}
            className={cx('modal', className)}
            transitionEnterTimeout={transitionEnterTimeout}
            transitionLeaveTimeout={transitionLeaveTimeout}
          >
            <FlexView>
              <BackgroundDimmer
                color='black'
                alpha={0.85}
                onClickOutside={dismissOnClickOut ? onDismiss : (() => {})}
                maxWidth={950}
              >
                <FlexView column grow className='modal-content'>
                  <FlexView
                    className='modal-header'
                    shrink={false}
                    vAlignContent='center'
                  >
                    <h3 className={cx('modal-title', titleProps.h3.className)}>
                      {title}
                    </h3>
                    {closeIcon && (
                      <Icon className='close-icon' icon='modal-times' onClick={onDismiss} />
                    )}
                  </FlexView>
                  <FlexView className='modal-body' column grow>
                    {children}
                  </FlexView>
                  <FlexView className='modal-footer' column shrink={false}>
                    {footer}
                  </FlexView>
                </FlexView>
              </BackgroundDimmer>
            </FlexView>
          </TransitionWrapper>
        ) : null}
      </React.addons.TransitionGroup>
    );
  }
}
