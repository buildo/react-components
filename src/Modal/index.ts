import { Modal } from './Modal';
export default Modal;
export { modalWithContext } from './modalWithContext';
