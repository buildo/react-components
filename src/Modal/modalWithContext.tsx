import * as React from 'react';
import { Modal } from './Modal';
import { ObjectOmit } from '../utils';

type ContextProps = {
  childContextTypes: React.ValidationMap<any>,
  getChildContext: () => object
};

type ContextWrapperProps = ObjectOmit<Modal.Props, keyof ContextProps>;

export const modalWithContext = (contextTypes: React.ValidationMap<any>) =>
  class ContextWrapper extends React.Component<ContextWrapperProps> {
    static contextTypes = contextTypes;

    render() {
      const contextProps: ContextProps = {
        childContextTypes: contextTypes,
        getChildContext: () => this.context
      };
      return <Modal {...this.props} {...contextProps} />;
    }
  };

