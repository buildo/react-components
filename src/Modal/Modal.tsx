import * as React from "react";
import * as cx from "classnames";
import omit = require("lodash/omit");
import { ModalPortal } from "./ModalPortal";
import FlexView from "react-flexview";
import { BackgroundDimmer } from "./BackgroundDimmer";

export type ModalDefaultProps = {
  /** whether the modal should be dismissed when clicking outside it */
  dismissOnClickOutside: boolean;
  /** called when modal is dismissed */
  onDismiss: (e: React.SyntheticEvent<HTMLDivElement>) => void;
  /** specify 'color' and 'alpha' for the overlay layer */
  overlay:
    | {
        color: string;
        alpha?: number;
      }
    | {
        color?: string;
        alpha: number;
      };
};

export type ModalRequiredProps = {
  /** modal content */
  children: JSX.Element;
  /** modal title */
  title?: string;
  /** modal footer */
  footer?: JSX.Element;
  /** close icon */
  iconClose?: JSX.Element;
  className?: string;
  style?: React.CSSProperties;
  id?: string;
};

export type ModalDefaultedProps = ModalPortal.Props &
  ModalRequiredProps &
  ModalDefaultProps;

export namespace Modal {
  export type Props = ModalPortal.Props &
    ModalRequiredProps &
    Partial<ModalDefaultProps>;
}

const LocalProps = {
  children: ReactChildren,
  title: t.maybe(t.String),
  footer: t.maybe(ReactChildren),
  iconClose: t.maybe(ReactChild),
  overlay: t.interface({
    color: t.maybe(t.String),
    alpha: t.maybe(t.Number)
  }),
  dismissOnClickOutside: t.maybe(t.Boolean),
  onDismiss: t.maybe(t.Function),
  className: t.maybe(t.String),
  style: t.maybe(t.Object),
  id: t.maybe(t.String)
};

export const Props = {
  ...PortalProps,
  ...LocalProps
};

@props(Props)
export class Modal extends React.Component<Modal.Props> {
  static defaultProps: ModalDefaultProps = {
    onDismiss: () => {},
    dismissOnClickOutside: true,
    overlay: {
      color: "black",
      alpha: 0.85
    }
  };

  render() {
    const {
      className,
      style,
      id,
      title,
      iconClose,
      footer,
      overlay,
      dismissOnClickOutside,
      onDismiss,
      children,
      ...props
    } = this.props as ModalDefaultedProps;
    const shouldRenderHeader = !!title || !!iconClose;
    const shouldRenderFooter = !!footer;

    const modalPortalProps: ModalPortal.Props = {
      ...omit(props, Object.keys(LocalProps)),
      className: cx("modal", className)
    } as ModalPortal.Props;

    return (
      <ModalPortal {...modalPortalProps}>
        <BackgroundDimmer
          {...{ ...overlay, style, id }}
          onClickOutside={dismissOnClickOutside ? onDismiss : undefined}
          maxWidth={950}
        >
          <FlexView className="modal-content" column grow>
            {shouldRenderHeader && (
              <FlexView
                className="modal-header"
                shrink={false}
                vAlignContent="center"
              >
                <FlexView className={cx("modal-title")} grow>
                  {title}
                </FlexView>
                {iconClose && (
                  <FlexView
                    className="modal-icon-close"
                    shrink={false}
                    marginLeft="auto"
                    onClick={onDismiss}
                  >
                    {iconClose}
                  </FlexView>
                )}
              </FlexView>
            )}
            <FlexView className="modal-body" column grow>
              {children}
            </FlexView>
            {shouldRenderFooter && (
              <FlexView className="modal-footer" column shrink={false}>
                {footer}
              </FlexView>
            )}
          </FlexView>
        </BackgroundDimmer>
      </ModalPortal>
    );
  }
}
