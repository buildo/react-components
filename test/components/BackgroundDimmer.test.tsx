import * as React from 'react';
import * as renderer from 'react-test-renderer';
import { shallow } from 'enzyme';

import { BackgroundDimmer } from '../../src/Modal/BackgroundDimmer';

const content = <div className="content">content</div>;

let consoleError: jest.SpyInstance<{}>;

beforeAll(() => {
  consoleError = jest.spyOn(console, 'error');
});

afterEach(() => {
  expect(consoleError).not.toHaveBeenCalled();
});

describe('BackgroundDimmer', () => {
  it('renders correctly', () => {
    const component = renderer.create(<BackgroundDimmer>{content}</BackgroundDimmer>);
    expect(component).toMatchSnapshot();
  });

  it('invokes the callback when the background is clicked', () => {
    const onClickOutside = jest.fn();
    const dimmer = shallow(
      <BackgroundDimmer onClickOutside={onClickOutside}>{content}</BackgroundDimmer>
    );
    dimmer.find('.main-content-wrapper').simulate('click');
    expect(onClickOutside).toBeCalled();
  });

  it("doesn't invoke the callback when the content is clicked", () => {
    const onClickOutside = jest.fn();
    const dimmer = shallow(
      <BackgroundDimmer onClickOutside={onClickOutside}>{content}</BackgroundDimmer>
    );
    dimmer.find('.content').simulate('click');
    expect(onClickOutside).not.toBeCalled();
  });
});
