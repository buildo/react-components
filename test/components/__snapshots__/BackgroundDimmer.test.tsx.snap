// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BackgroundDimmer renders correctly 1`] = `
<div
  className="background-dimmer"
>
  <div
    style={
      Object {
        "backgroundColor": "black",
        "bottom": 0,
        "left": 0,
        "opacity": 0.5,
        "position": "fixed",
        "right": 0,
        "top": 0,
        "zIndex": 99999,
      }
    }
  />
  <div
    className="main-content-wrapper"
    onClick={[Function]}
    onTouchMove={[Function]}
    onWheel={[Function]}
    style={
      Object {
        "alignItems": "center",
        "bottom": 0,
        "boxSizing": "border-box",
        "display": "flex",
        "flex": "0 1 auto",
        "flexDirection": "row",
        "flexWrap": "nowrap",
        "height": undefined,
        "justifyContent": "center",
        "left": 0,
        "marginBottom": undefined,
        "marginLeft": undefined,
        "marginRight": undefined,
        "marginTop": undefined,
        "minHeight": 0,
        "minWidth": 0,
        "position": "fixed",
        "right": 0,
        "top": 0,
        "width": undefined,
        "zIndex": 100000,
      }
    }
  >
    <div
      className="centered-content-wrapper"
      onClick={[Function]}
      style={
        Object {
          "alignItems": undefined,
          "boxSizing": "border-box",
          "display": "flex",
          "flex": "0 1 auto",
          "flexDirection": "column",
          "flexWrap": "nowrap",
          "height": "auto",
          "justifyContent": undefined,
          "marginBottom": undefined,
          "marginLeft": undefined,
          "marginRight": undefined,
          "marginTop": undefined,
          "maxHeight": "90%",
          "maxWidth": "90%",
          "minHeight": 0,
          "minWidth": 0,
          "width": "auto",
        }
      }
    >
      <div
        className="content"
      >
        content
      </div>
    </div>
  </div>
</div>
`;
